File: rules-interpreter/src/main/java/io/github/iamrenny/ruleflow/visitors/ActionsVisitor.java
package io.github.iamrenny.ruleflow.visitors;

import io.github.iamrenny.ruleflow.RuleFlowLanguageParser;
import io.github.iamrenny.ruleflow.utils.Pair;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class ActionsVisitor {

    private final Visitor visitor;

    public ActionsVisitor(Visitor visitor) {
        this.visitor = visitor;
    }

    public List<Pair<String, Map<String, String>>> visit(RuleFlowLanguageParser.ActionsContext context) {
        return context.action().stream()
            .map(action -> {
            String name;
            if (action.K_ACTION() != null) {
                name = action.param_value.getText().replace("'", "");
            } else if ("manual_review".equals(action.action_id.getText())) {
                name = action.action_id.getText().replace("'", "");
            } else if ("apply_restriction".equals(action.action_id.getText())) {
                name = action.action_id.getText().replace("'", "");
            } else if (action.action_id.getText() != null) {
                name = action.action_id.getText().replace("'", "");
            } else {
                throw new IllegalArgumentException("Cannot find action name or identifier in " + action.getText());
            }
            return new Pair<>(name, new ActionVisitor(visitor).visit(action));
        })
        .collect(Collectors.toList());
    }
}

class ActionVisitor {

    private final Visitor visitor;

    public ActionVisitor(Visitor visitor) {
        this.visitor = visitor;
    }

    public Map<String, String> visit(RuleFlowLanguageParser.ActionContext action) {
        if (action.action_params() != null) {
            return action.action_params().param_pairs().param_pair().stream()
                .collect(Collectors.toMap(
                    a -> a.field_name.getText().replace("'", ""),
                    a -> evaluateActionParamValue(a.field_value)
            ));
        } else {
            return Map.of();
        }
    }

    private String evaluateActionParamValue(RuleFlowLanguageParser.ActionParamValueContext fieldValue) {
        if (fieldValue.validValue() != null) {
            // Handle string literals, numbers, booleans, etc.
            return fieldValue.validValue().getText().replace("'", "");
        } else if (fieldValue.validProperty() != null) {
            // Handle property references - resolve the actual value from the request data
            try {
                Object resolvedValue = visitor.visit(fieldValue.validProperty());
                return resolvedValue != null ? resolvedValue.toString() : "null";
            } catch (Exception e) {
                // If property resolution fails, return the property name as fallback
                return fieldValue.validProperty().getText();
            }
        } else {
            throw new IllegalArgumentException("Unsupported action parameter value type: " + fieldValue.getText());
        }
    }
}

File: rules-interpreter/src/main/kotlin/com/rappi/analang/visitors/ActionsVisitor.kt
package com.rappi.analang.visitors

import com.rappi.fraud.analang.ANAParser
import kotlin.streams.toList

class ActionsVisitor {
    fun visit(context: ANAParser.ActionsContext): List<Pair<String, Map<String, String>>> {
        return context.action()
            .stream()
            .map { action ->
                val name = when {
                    action.K_ACTION() != null -> action.param_value.text.replace("'", "")
                    action.action_id.text == "manual_review" -> action.action_id.text.replace("'", "")
                    action.action_id.text == "apply_restriction" -> action.action_id.text.replace("'", "")
                    else -> error("Cannot find action name or identifier in ${action.text}")
                }
                Pair(name, ActionVisitor().visit(action))
            }
            .toList()
    }
}

class ActionVisitor {

    fun visit(action: ANAParser.ActionContext): Map<String, String> {

        return if (action.action_params() != null)
            action.action_params().param_pairs().param_pair()
                .map { a ->
                    Pair(a.field_name.text.replace("'",""), a.field_value.text.replace("'",""))
                }
                .toMap()
        else mapOf()
    }
}
